### ISSUE TITLE
Tactic stuck due to bad memoization

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1271]
The following fails:

```
module Minimal

open FStar.Tactics

#set-options "--ugly --print_full_names --print_bound_var_types"

let test =
  assert_by_tactic (True ==> True)
    (b <-- implies_intro;
     qb <-- quote b;
     qf <-- quote (fun (b: binder) -> print "f"); // f: tactic unit
     let q_fofb = mk_app qf [(qb, Q_Explicit)] in
     print ("::: " ^ term_to_string q_fofb);;
     unquote #(tactic unit) q_fofb;;
     fail "A")
```

It says `Failed to resolve implicit argument of type 'FStar.Reflection.Types.binder' introduced in (?410 uu___#713907)`

I chatted about this briefly with @mtzguido; CC @nikswamy.

**Comment by mtzguido** [comment]
This now works in my branch. The problem is that we were not able to typecheck the embedded alien terms again, and since they are just a Tm_unknown + meta, you were getting this error. The fix is to carry their type and use it when typechecking. There's also a fix needed in the normalizer, but I'm not super convinced of it, so not merging for now.

**Comment by mtzguido** [comment]
A partial fix now in master. Your example works, and you can unquote the tactic, but if you try to call it the normalizer will get stuck. I think the memoization logic is to blame, but fixing it made CI time rise by 50%; that's the part I'm dubious about, so that did not go into master yet.

**Comment by mtzguido** [comment]
This is the test
```
module Aliens

open FStar.Tactics

(* Testing that aliens are typechecked/unquoted properly *)

let test =
  assert_by_tactic (True ==> True)
    (b <-- implies_intro;
     qb <-- quote b;
     qf <-- quote (fun (b: binder) -> print "f"); // f: tactic unit
     let q_fofb = mk_app qf [(qb, Q_Explicit)] in
     print ("::: " ^ term_to_string q_fofb);;
     tac <-- unquote #(tactic unit) q_fofb;
     tac;;
     trivial)
```
which fails with
```
Failure("Expected Result.Success or Result.Failed applied to a single argument, got (match
    match reify ((fun b -> FStar.Tactics.Builtins.print \"f\") _ ()) _ with
    | FStar.Tactics.Result.Success #uu___566985 (#uu___566981, #uu___566982, #a, #q) ->
      (let () = FStar.Tactics.Types.tracepoint q in
        reify (FStar.Tactics.Builtins.trivial ()) (FStar.Tactics.Types.decr_depth q))
      <:
      FStar.Tactics.Result.__result Prims.unit
    | FStar.Tactics.Result.Failed #uu___567613 (#uu___567609, #uu___567610, #msg, #q) ->
      FStar.Tactics.Result.Failed (FStar.Pervasives.Native.Mktuple2 msg q)
  with
  | FStar.Tactics.Result.Success #uu___566985 (#uu___566981, #uu___566982, #a, #q) ->
    (let () = FStar.Tactics.Types.tracepoint q in
      reify (FStar.Tactics.Effect.return a ()) (FStar.Tactics.Types.decr_depth q))
    <:
    FStar.Tactics.Result.__result Prims.unit
  | FStar.Tactics.Result.Failed #uu___567613 (#uu___567609, #uu___567610, #msg, #q) ->
    FStar.Tactics.Result.Failed (FStar.Pervasives.Native.Mktuple2 msg q))
<:
FStar.Tactics.Result.__result Prims.unit")
```
The commit to fix memoization is 7e32a4e59785d51b78a38210f9440211a4c59e8f, but seems to make things much slower (which I think is a symptom that there are many examples where we're memoizing not-yet-normal-forms).

**Comment by mtzguido** [comment]
Amazingly, this is also fixed by other normalizer and reification fixes. The programming gods are smiling upon us. Merging it soon, so closing this.
